presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java

```diff
@@ -0,0 +1,308 @@
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(queryId, source);
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .addValue(queryId)
+                    .addValue(source)
+                    .toString();
+        }
+    }
+
+    @ThreadSafe
+    private static class DynamicFilterSummaryWithSenders
+    {
+        private final Map<StageTaskKey, SenderStats> senderStats = new HashMap<>();
+        private Set<StageTaskKey> registeredTasks = ImmutableSet.of();
+        private DynamicFilterSummary dynamicFilterSummary;
+
+        synchronized void registerTasks(Set<StageTaskKey> tasks)
+        {
+            this.registeredTasks = ImmutableSet.copyOf(tasks);
+        }
+
+        synchronized Optional<DynamicFilterSummary> getSummaryIfReady()
+        {
+            if (registeredTasks.isEmpty() || !senderStats.keySet().equals(registeredTasks)) {
+                return Optional.empty();
+            }
+
+            for (SenderStats stageTaskStats : senderStats.values()) {
+                if (!stageTaskStats.isCompleted()) {
+                    return Optional.empty();
+                }
+            }
+            // All drivers of all tasks on build side have completed and reported dynamic filters
+            return Optional.of(dynamicFilterSummary);
+        }
+
+        synchronized DynamicFilterSummary addSummary(StageTaskKey stageTaskId, int driverId, DynamicFilterSummary summary, int expectedSummariesCount)
+        {
+            SenderStats stats = senderStats.computeIfAbsent(stageTaskId, ignored -> new SenderStats());
+            stats.addSummary(driverId, expectedSummariesCount);
+
+            if (dynamicFilterSummary == null || summary.getTupleDomain().isAll()) {
+                dynamicFilterSummary = summary;
+            }
+            else {
+                dynamicFilterSummary = dynamicFilterSummary.mergeWith(summary);
+            }
+
+            if (stats.isCompleted()) {
+                return dynamicFilterSummary;
+            }
+
+            return null;
+        }
+
+        @NotThreadSafe
+        private static class SenderStats
+        {
+            // collection of driverIDs reported
+            private final Set<Integer> driverIDs = new HashSet<>();
+            private int expectedSummariesCount = -1;
+
+            int getExpectedSummariesCount()
+            {
+                return expectedSummariesCount;
+            }
+
+            public boolean isCompleted()
+            {
+                return (expectedSummariesCount != -1) && (driverIDs.size() == expectedSummariesCount);
+            }
+
+            void addSummary(Integer driverId, int expectedSummariesCount)
+            {
+                if (driverIDs.contains(driverId)) {
+                    // skip existing driver IDs in case of HTTP retries
+                    return;
```

Comment:
I've simplified the implementation to now merge driver summaries within the worker node so that we don't need to send driverId and expectedSummariesCount to coordinator